pkg/update/update.go

```diff
@@ -65,6 +65,60 @@ func (c *updater) WithStatusWriter(writer statuswriter.StatusWriter) *updater {
 	return c
 }
 
+func (c *updater) PrepareForVersion(
+	ctx context.Context,
+	pkgName string,
+	pkgVersion string) (*UpdateTransaction, error) {
+	c.status.Start()
+	defer c.status.Stop()
+	c.status.SetStatus("Collecting installed package")
+
+	var packagesToUpdate []v1alpha1.Package
+	var pkg v1alpha1.Package
+	if err := c.client.Packages().Get(ctx, pkgName, &pkg); err != nil {
+		return nil, fmt.Errorf("failed to get package %v: %v", pkgName, err)
+	}
+	packagesToUpdate = append(packagesToUpdate, pkg)
+
+	c.status.SetStatus("Updating package index")
+	var index repo.PackageRepoIndex
+	if err := repo.FetchPackageRepoIndex("", &index); err != nil {
+		return nil, fmt.Errorf("failed to fetch index: %v", err)
+	}
+
+	requirementsSet := make(map[dependency.Requirement]struct{})
+	var tx UpdateTransaction
+	for _, pkg := range packagesToUpdate {
+		for _, indexItem := range index.Packages {
+			if indexItem.Name == pkg.Name {
+				item := updateTransactionItem{Package: pkg, Version: pkgVersion}
```

Comment:
Please have a look at the similar function `Prepare`: At this point, it is important to check whether the requested version is newer than the currently installed version. Downgrade should not be possible.

Reply:
Can you please guide me on this. According to me we should use the `semver.IsVersionUpgradable` function.

Reply:
That's exactly the way. Use this function, and if it returns true, you can continue, otherwise you can return because the update is not possible.

Reply:
Well I tried using this code:
```
parsedLatest, err := s.NewVersion(indexItem.LatestVersion)
			if err != nil {
				fmt.Fprintf(os.Stderr, "Error in creating the parsed version\n")
			}
			parsedDesired, err := s.NewVersion(pkgVersion)
			if err != nil {
				fmt.Fprintf(os.Stderr, "Error in creating the parsed version\n")
			}
			if !semver.IsVersionUpgradable(parsedLatest, parsedDesired) {
				fmt.Fprintf(os.Stderr, "Can't update downgraded version\n")
				return nil, nil
			}
```

But here when I am updating from the upper-version to the latest version it is stopping that

Reply:
What is upper version and what is latest version in your scenario?

Reply:
Upper version: argo-cd v2.10.5+1
Latest Version: argo-cd v2.10.4+1

Reply:
Ok, and which one is currently installed?

Reply:
Upper Version

Reply:
Then it works exactly as intended. 

You have argo-cd v2.10.5+1 installed, and argo-cd v2.10.4+1 is of course older than your currently installed version. This means you cannot update to v2.10.4+1.

Reply:
Ya but v2.10.4+1 is the latest version right? and users can update to the latest version.

Reply:
Sorry for the confusion about the word "latest", but that's just where we are right now. 

Latest version in our context means, it is the version which is "marked as latest" in the package repository. See for example: https://github.com/glasskube/packages/blob/main/packages/argo-cd/versions.yaml . Here, 2.10.4+1 is marked as latest, but clearly 2.10.5+1 is newer according to semantic versioning. 

For your feature, it does not matter at all what is marked as latest version. The only thing that matters is, you cannot "update" to a version which is smaller than the currently installed version.

Reply:
Ohk understood will do this test and then commit the code:
https://github.com/glasskube/glasskube/pull/461#pullrequestreview-1978859558